kart/log.py

```diff
@@ -72,65 +118,110 @@ def parse_extra_args(
-                    RemovalInKart012Warning,
-                )
-                other_args.append(arg)
-                continue
-
-            range_parts = re.split(r"\.\.\.?", arg)
-            if len(range_parts) > 2:
-                # not a valid range or ref, must be a path
-                # Treat remaining args as paths
-                paths = args[i:]
-                break
 
-            try:
-                for part in range_parts:
-                    repo.resolve_refish(part or "HEAD")
-            except (KeyError, pygit2.InvalidSpecError):
-                # not a commit-ish.
-                # Treat remaining args as paths
-                paths = args[i:]
-                break
-            else:
-                other_args.append(arg)
+    options = []
+    commits = []
+    allow_paths = dash_index is None
+    for arg in args:
+        arg_type = get_arg_type(repo, arg, allow_paths=allow_paths)
+        {
+            LogArgType.OPTION: options,
+            LogArgType.COMMIT: commits,
+            LogArgType.PATH: paths,
+        }[arg_type].append(arg)
 
     if max_count is not None:
-        other_args.append(f"--max-count={max_count}")
+        options.append(f"--max-count={max_count}")
     if skip is not None:
-        other_args.append(f"--skip={skip}")
+        options.append(f"--skip={skip}")
     if since is not None:
-        other_args.append(f"--since={since}")
+        options.append(f"--since={since}")
     if until is not None:
-        other_args.append(f"--until={until}")
+        options.append(f"--until={until}")
     # These ones can be specified more than once
     if author:
-        other_args.extend(f"--author={a}" for a in author)
+        options.extend(f"--author={a}" for a in author)
     if committer:
-        other_args.extend(f"--committer={c}" for c in committer)
+        options.extend(f"--committer={c}" for c in committer)
     if grep:
-        other_args.extend(f"--grep={g}" for g in grep)
-    return other_args, list(paths)
+        options.extend(f"--grep={g}" for g in grep)
+    return options, commits, paths
+
+
+def find_dataset(ds_path, repo, commits):
+    """Finds a dataset by name, so long as it is found somewhere in the given commits / refs / ranges."""
+    if ds_path in repo.datasets():
+        return repo.datasets()[ds_path]
+    cmd = [
+        "git",
+        "-C",
+        repo.path,
+        "log",
+        "--max-count=1",
+        "--format=%H",
+        *commits,
+        "--",
+        ds_path,
+    ]
+    try:
+        r = subprocess.run(
+            cmd,
+            encoding="utf8",
+            check=True,
+            capture_output=True,
+            env=tool_environment(),
+        )
+    except subprocess.CalledProcessError as e:
+        raise SubprocessError(
+            f"There was a problem with git log: {e}", called_process_error=e
+        )
+    if r.stdout:
+        try:
+            return repo.datasets(r.stdout.strip())[ds_path]
```

Comment:
The log command above will find the last commit in which the given dataset was modified.

If that commit actually deleted the dataset, I think this will raise a KeyError (which is ignored here), and we'll return None, which is handled as if the dataset didn't exist in the entire log.

Instead we should probably use the dataset from the parent commit in that case.

Reply:
Handled this case and added a test.